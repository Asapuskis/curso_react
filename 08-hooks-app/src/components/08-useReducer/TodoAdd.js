import React from 'react'
import { useForm } from '../../hooks/useForm'

export const TodoAdd = ( { handleAddTodo }) => {

    // Utilizamos nuestro custom hook useForm para manejar los formularios. Iniciandolo a vacío.
    const { formValues, handleInputChange, reset } = useForm({
        description: ""
    })

    // Creamos una función para manejar el submit del formulario
    const handleSubmit = (e) => {
        // preventDefault evita que se recargue por completo la página.
        e.preventDefault();

        // Validamos que la nueva tarea sea más larga que un caracter para evitar meter mierda
        if (formValues.description.trim().length <= 1)
            return;

        // Generamos el nuevo objeto TODO con la información recogida del formulario.
        const newTodo = {
            id: new Date().getTime(),
            desc: formValues.description,
            done: false
        };

        // Mediante la función dispatch, le pasamos el objeto que queremos, en este caso, añadir a la lista.
        handleAddTodo( newTodo );

        // Ejecutamos reset para limpiar la caja de texto una vez se haya introducido la nueva tarea a la lista.
        reset();
    }

    return (
    <>
        <h4>Agregar Todos</h4>
        <hr/>
        <form onSubmit={handleSubmit}>
            <input
                type='text'
                name='description'
                placeholder='Aprender...'
                autoComplete='off'
                className='form-control'
                value={ formValues.description }
                onChange={ handleInputChange }
            />
            <button
                type='submit'
                className='btn btn-outline-primary mt-1 btn-block form-control'
            >
                Agregar
            </button>
        </form>
    </>
  )
}
