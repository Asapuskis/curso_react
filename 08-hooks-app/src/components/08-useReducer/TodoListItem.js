import React from 'react'

export const TodoListItem = ( { todo, i, handleDelete, handleToggle }) => {
  return (
    <li key={ todo.id } className="list-group-item">
        <p
            className={ `${ todo.done && 'complete' }` }
            onClick={ () => handleToggle(todo.id)}
        >
            {i+1}. { todo.desc }
        </p>
        <button
            className='btn btn-danger'
            onClick={ () => handleDelete(todo.id)}>
            Borrar
        </button>
    </li>
  )
}
