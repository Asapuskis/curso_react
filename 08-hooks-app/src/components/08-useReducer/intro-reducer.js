// Introducción del concepto reducer
// Un Reducer sirve para gestionar los estados de una aplicación de manera controlada y segura.
// Por lo general, un reducer va a tener las funcionalidades insertar, actualizar y borrar.
// Estas funciones serán las que se ejecuten sobre el estado y lo modificarán.

console.log('kps');

// Creamos una initialTodo para la prueba, este va a ser el estado inicial
const initialTodo = [{
    id: 1,
    todo: 'Comprar leche',
    done: false
}];

// Función reducer. Recibe dos parámetros, el estado y la acción que hay que realizar sobre esos estados
const todoReducer = ( state = initialTodo, action ) => {

    // Si la acción es de tipo agregar, añadimos a la lista de estados el nuevo estado.
    if ( action?.type  === 'add' ) {
        return [...state, action.payload ];
    }


    return state;
}

let todoList = todoReducer();

const newTodo = {
    id: 2,
    todo: 'Comprar Sanguis',
    done: false
};

// Por defecto, el objeto action va a tener dos atributos
// 1. type: Para indicar qué acción se va a realizar en el reducer 'add', 'update', 'delete'
// 2. payload: El nuevo estado que se va a introducir, modificar, etc.
const addAction = {
    type: 'add',
    payload: newTodo
}

todoList = todoReducer( todoList, addAction)

console.log(todoList);