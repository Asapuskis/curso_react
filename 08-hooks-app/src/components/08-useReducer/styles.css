body {
    padding: 50px;
}

.list-group-item {
    align-items: center;
    display: flex;
    justify-content: space-between;
}

.complete {
    text-decoration: line-through;
}

p {
    cursor: pointer;
    margin: 0px;
}