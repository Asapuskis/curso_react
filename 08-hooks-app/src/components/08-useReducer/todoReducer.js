// Funcion para controlar el reducer, utilizaremos un switch.
export const todoReducer = ( state = [], action ) => {
    switch ( action.type ) {
        // Action add: AÃ±adimos un nuevo state a la lista de state
        case 'add':
            return [ ...state, action.payload ];
        // Action delete: Devolvemos con filter todos los elementos que no coincidan con el action.payload.id
        case 'delete':
            return state.filter( state => state.id !== action.payload.id );
        case 'toggle':
            return state.map( todo => {
                if ( todo.id === action.payload.id ) {
                    return {
                        ...todo,
                        done: !todo.done
                    }
                } else {
                    return todo;
                }
            });
        default:
            return state;
    }
}