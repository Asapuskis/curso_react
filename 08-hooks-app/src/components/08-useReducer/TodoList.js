import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodoList = ( { todoList, handleDelete, handleToggle } ) => {
  return (
    <ul className='list-group list-group-flush'>
        {
            todoList.map( (todo, i) => (
                <TodoListItem
                    key={ todo.id }
                    todo={todo}
                    i={i}
                    handleDelete={handleDelete}
                    handleToggle={handleToggle}
                />
            ))
        }
    </ul>
  )
}
