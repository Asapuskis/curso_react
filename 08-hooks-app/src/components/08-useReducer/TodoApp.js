import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';

import './styles.css';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

// En este componente se usará el hook useReducer
export const TodoApp = () => {

    // useReducer tiene 3 parámetros:
    // const [state, dispatch] = useReducer(reducer, initialArg, init);
    // 1. Función reducer
    // 2. Estado inicial
    // 3. (optional) Función para inicializar el state de manera diferida: https://es.reactjs.org/docs/hooks-reference.html#lazy-initialization
    // Devuelve un array con dos elementos
    // 1. El estado actualizado
    // 2. (opcional) La función dispatch que será para editar el estado de manera semejante al useState

    const init = () => {

        // Inicializamos la lista a vacío o lo que haya en localStorage. getItem devuelve null si no se encuentra, por lo tanto, si no se encuentra
        // nada en 'todos', devolverá por defecto la lista vacía []
        return JSON.parse(localStorage.getItem('todos')) || [];

    };

    // Inicializamos la lista con useReducer y el estado inicial. Utilizando la función init, inicializamos la lista de manera más eficiente
    const [ todoList, dispatch ] = useReducer(todoReducer, [], init)

    // Creamos un useEffect para actualizar el localStorage con los valores de la lista al modificarse.
    // De esta manera, tendremos los datos persistentes y accesibles aunque se recarge el navegador.
    useEffect( () => {
        localStorage.setItem('todos', JSON.stringify(todoList));
    }, [todoList]);

    // Función para manejar el borrado de tareas de la lista.
    const handleDelete = ( todoId ) => {
        console.log(todoId);
        const deleteTodo = {
            id: todoId
        };
        const deleteAction = {
            type: 'delete',
            payload: deleteTodo
        };
        dispatch( deleteAction );
    }

    const handleToggle = ( todoId ) => {
        const toggleTodo = {
            id: todoId
        };
        const toggleAction = {
            type: 'toggle',
            payload: toggleTodo
        };
        dispatch(toggleAction);
    }

    const handleAddTodo = ( newTodo ) => {
        // Generamos el nuevo objeto action para pasarselo al useReducer
        const addAction = {
            type: 'add',
            payload: newTodo
        };

        dispatch(addAction);
    }


    return (
        <div>
            <h1>TodoApp <small>{todoList.length}</small></h1>
            <hr/>
            <div className='row'>
                <div className='col-7'>
                    <TodoList
                        todoList={todoList}
                        handleDelete={handleDelete}
                        handleToggle={handleToggle}
                    />
                </div>
                <div className='col-5'>
                    <TodoAdd
                        handleAddTodo={handleAddTodo}
                    />
                </div>
            </div>
        </div>
    )
}
