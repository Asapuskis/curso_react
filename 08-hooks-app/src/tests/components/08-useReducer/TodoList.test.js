import { shallow } from 'enzyme';
import { TodoList } from '../../../components/08-useReducer/TodoList';
import { demoTodos } from '../../fixtures/demoTodos';


describe('Pruebas de <TodoList />', () => {

    const handleDelete = jest.fn();
    const handleToggle = jest.fn();

    const wrapper = shallow(
        <TodoList
            todoList={demoTodos}
            handleDelete={handleDelete}
            handleToggle={handleToggle}
        /> );

    test('Debe mostrarse un snapshot correctamente', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('Debe tener dos elementos <TodoListItems />', () => {

        expect(wrapper.find('TodoListItem').length).toBe(demoTodos.length);
        expect(wrapper.find('TodoListItem').at(0).prop('handleDelete')).toEqual(expect.any(Function));

    });

})