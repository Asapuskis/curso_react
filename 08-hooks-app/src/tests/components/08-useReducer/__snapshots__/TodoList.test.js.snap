// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de <TodoList /> Debe mostrarse un snapshot correctamente 1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    i={0}
    key="1"
    todo={
      Object {
        "desc": "Aprender React",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    i={1}
    key="2"
    todo={
      Object {
        "desc": "Aprender Mongo",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
